sycl/include/sycl/detail/util.hpp

```diff
@@ -67,6 +69,30 @@ struct CmpCStr {
 
 using SerializedObj = std::vector<unsigned char>;
 
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
+// We need special handling of std::string to handle ABI incompatibility
+// for get_info<>() when it returns std::string and vector<std::string>.
+// For this purpose, get_info_impl<>() is created to handle special
+// cases, and it is only called internally and not exposed to the user.
+// The following ReturnType structure is intended for general return type,
+// and special return types (std::string and vector of it).
+template <typename T> struct ABINeutralT { using type = T; };
+
+template <> struct ABINeutralT<std::string> { using type = detail::string; };
+
+template <> struct ABINeutralT<std::vector<std::string>> {
+  using type = std::vector<detail::string>;
+};
+
+template <> struct ABINeutralT<detail::string> { using type = std::string; };
+
+template <> struct ABINeutralT<std::vector<detail::string>> {
+  using type = std::vector<std::string>;
+};
+
+template <typename T> using ABINeutralT_t = typename ABINeutralT<T>::type;
```

Comment:
This is wrong. Should have been either a separate trait or, preferably, something like this

```
auto foo(...) {
  if constexpr (/* detail::string */) {
    return std::string{...};
  } else {
  }
}
```

Reply:
Are you suggesting to replace `ABINeutralT `with a function?
`ABINeutralT `is a polymorphic type that I needed to represent `std::string`, `std::vector<std::string>`, others when I register `get_info_impl<>` for all various properties in device.cpp and platform.cpp.
Your requested function foo() is defined in `convert_to/from_abi_neutral()` in platform.hpp.

Reply:
This hasn't been addressed.

Reply:
Could you answer my question above? I am not sure if I understand your request.
Are you asking to remove ABINetraulT and replace with a function that does conversion?
I am not sure how it is going to solve/improve our issues.

Reply:
I'm just saying that the code you have currently is poor. You can't name something "ABINeutral" and return `std::string` that we're trying to eliminate from the ABI boundary.

Reply:
I just removed lines 87-91 since they are not needed anymore. :)
Hopefully, this resolved your concern.